warehouse/query-core/src/main/java/datawave/query/jexl/visitors/UniqueExpressionTermsVisitor.java

```diff
@@ -0,0 +1,156 @@
+    
+    private static final Logger log = Logger.getLogger(UniqueExpressionTermsVisitor.class);
+    
+    /**
+     * Apply this visitor until the query tree stops changing or the maximum number of iterations is exceeded.
+     *
+     * @param node
+     *            - the root node for a query tree
+     * @param <T>
+     * @return
+     */
+    public static <T extends JexlNode> T enforce(T node, int maxIterations) {
+        if (node == null)
+            return null;
+        
+        // Operate on copy of query tree
+        T copy = (T) copy(node);
+        
+        UniqueExpressionTermsVisitor visitor = new UniqueExpressionTermsVisitor();
+        boolean changesMade = true;
+        
+        while (visitor.iterations < maxIterations && changesMade) {
+            visitor.iterations++;
+            
+            copy = TreeFlatteningRebuildingVisitor.flatten(copy);
+            copy.jjtAccept(visitor, null);
+            
+            // Changes were made if the two trees are not equals
+            changesMade = !TreeEqualityVisitor.isEqual((ASTJexlScript) node, (ASTJexlScript) copy, new TreeEqualityVisitor.Reason());
+            node = copy;
+        }
+        
+        // Log if we stopped visiting due to exceeding the max iterations
+        if (visitor.iterations >= maxIterations && visitor.iterations > 0) {
+            if (log.isDebugEnabled()) {
+                log.debug("UniqueExpressionTermsVisitor exceeded maximum iteration (" + maxIterations + ").");
+            }
+        }
+        
+        if (log.isDebugEnabled()) {
+            log.debug("UniqueExpressionTermsVisitor took " + visitor.iterations + " iterations to fully process query tree.");
+            log.debug("UniqueExpressionTermsVisitor removed " + visitor.duplicates + " duplicate terms");
+        }
+        return node;
+    }
+    
+    @Override
+    public Object visit(ASTOrNode orNode, Object data) {
+        
+        // Post-order traversal
+        orNode.childrenAccept(this, data);
+        
+        List<JexlNode> children = removeDuplicateChildNodes(orNode);
+        
+        if (children.size() == 1) {
+            // If only one child remains, swap child node for current node.
+            JexlNode child = children.get(0);
+            replaceChild(orNode.jjtGetParent(), orNode, child);
+            return child;
+        } else {
+            // If two or more children remain, replace current children with new children
+            ASTOrNode copy = new ASTOrNode(ParserTreeConstants.JJTORNODE);
+            copy.image = orNode.image;
+            copy.jjtSetParent(orNode.jjtGetParent());
+            for (int ii = 0; ii < children.size(); ii++) {
+                copy.jjtAddChild(children.get(ii), ii);
+            }
+            replaceChild(orNode.jjtGetParent(), orNode, copy);
+            orNode = copy;
+        }
+        return orNode;
+    }
+    
+    @Override
+    public Object visit(ASTAndNode andNode, Object data) {
```

Comment:
the code here is nearly identical to visit(ASTOrNode, ...) refactor out a method and reuse

Reply:
Refactored in next commit.